### PR TITLE
Separate kwargs of tokenizer and feature_extractor in `ClapProcessor`

### PR DESCRIPTION
**Pull request by anmolojha**
# What does this PR do?
Currently, `ClapProcessor` shares kwargs between the tokenizer and the feature extractor. This PR introduces separate kwargs for both of them.

This was discussed in [comments](https://github.com/huggingface/transformers/issues/23648#issuecomment-1557532041) of #23648. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sanchit-gandhi

### CODE REVIEW

File: src/transformers/models/clap/processing_clap.py [hunk 1]
```diff
@@ -39,13 +39,21 @@ class ClapProcessor(ProcessorMixin):
...
 
-    def __call__(self, text=None, audios=None, return_tensors=None, **kwargs):
+    def __call__(
+        self,
+        text=None,
+        audios=None,
+        return_tensors=None,
+        tokenizer_kwargs=None,
+        feature_extractor_kwargs=None,
+        **kwargs,
```
Comment (sanchit-gandhi) [inline src/transformers/models/clap/processing_clap.py:49]: I wonder whether a better API is one where we prepend the tokenizer kwargs with `tokenizer_` and the feature extractor kwargs with `feature_extractor_`, e.g.:
```python
processor(text="the cat sat on the mat", audio=sample["array"], tokenizer_padding=True, feature_extractor_padding=False)
```

So we only have one argument `kwargs`, but we separate these into the `feature_extractor_kwargs` and the `tokenizer_kwargs` based on the naming of the kwargs. E.g. `feature_extractor_padding` becomes the kwarg `padding` for the feature extractor, and `tokenizer_padding` becomes the kwarg `padding` for the tokenizer.

What we can do to extract the relevant tokenizer/feature extractor kwargs would look very similar to how we extract the encoder/decoder kwargs in the `EncoderDecoderModel`:

https://github.com/huggingface/transformers/blob/33b5ef5cdf403ae42f7cdddf906e39073583f062/src/transformers/models/encoder_decoder/modeling_encoder_decoder.py#L458-L464

Comment (anmolojha) [reply src/transformers/models/clap/processing_clap.py:49]: Hey, sorry for the delayed response 🙇 

>could we add a test to verify we get the correct functionality in [test_processor_clap.py](https://github.com/huggingface/transformers/blob/main/tests/models/clap/test_processor_clap.py) please?

Sure, will do that.

>I wonder whether a better API is one where we prepend the tokenizer kwargs with tokenizer_ and the feature extractor kwargs with feature_extractor_

Yes, that will work too, and the arguments of `processor` look cleaner!

I still like the current approach though because:
1. I think it is more explicit. Argument names `tokenizer_kwargs` and `feature_extractor_kwargs` hints that something is different with `kwargs`. The proposed approach on the other hand relies more on people reading the documentation.
2. I don't have a lot of experience but I don't get the motivation behind adding a prefix. If I wanted to keep only one argument, I am not sure why shouldn't I use a nested dictionary.
```python
kwargs = {
    "encoder_kwargs": {
        "padding": True
    },
    "decoder_kwargs": {
        "padding": False
    },
    "return_tensor": "pt",
}
``` 

But I understand that consistency is very important. If there are multiple instances where the proposed approach is used, we should go with that. Let me know what you think.

**Comment by anmolojha** [comment]
@sanchit-gandhi can we fix both the points discussed in the issue (#23648) (shared kwargs and unexpected padding behaviour) in the same PR or should we have separate PRs?

File: src/transformers/models/clap/processing_clap.py (hunk 1 above)
Comment (sanchit-gandhi) [reply src/transformers/models/clap/processing_clap.py:49]: Hey @anmolojha - no need to apologise! Very happy you've come back to discuss this issue 🤗

The kwargs won't actually be a nested dictionary, but rather a flat dictionary with names prepended at the start, e.g.:

```python
kwargs = {
    "feature_extractor_padding": True,
    "tokenizer_padding": True,
    "tokenizer_return_tensors": "pt",
}
```

Then in the processor code we filter all feature extractor args (if they're prepended with `feature_extractor_`) and tokenizer args (if they're prepended with `tokenizer_`)

Would you be more inclined for a design like this? IMO this is more consistent with the rest of the lib and probably quite straightforward to adopt to as a user! Let me know what you think!

**Comment by sanchit-gandhi** [comment]
Separate PR would be preferable since you can work on each bit in isolation (making it easier for yourself and the reviewer) and merge this one as soon as you have it ready :)

File: src/transformers/models/clap/processing_clap.py (hunk 1 above)
Comment (sanchit-gandhi) [reply src/transformers/models/clap/processing_clap.py:49]: LMK @anmolojha if you have any questions regarding the proposed design! I think the one proposed in my previous comment is most in-keeping with the rest of the `transformers` library, and it shouldn't be too difficult to adapt the current code to this format. But keen to hear your thoughts and whether you'd do anything differently!

**Comment by sanchit-gandhi** [comment]
Think it's going to be quite fast to finish this PR - WDYT @anmolojha?

**Comment by anmolojha** [comment]
Hey @sanchit-gandhi, thanks for following up 🙏
This got delayed because of a bunch of reasons, I will close this this weekend.

**Comment by sanchit-gandhi** [comment]
Awesome! Looking forward to it @anmolojha!

File: src/transformers/models/clap/processing_clap.py [hunk 2]
```diff
@@ -56,15 +63,19 @@ def __call__(self, text=None, audios=None, return_tensors=None, **kwargs):
...
             return_tensors (`str` or [`~utils.TensorType`], *optional*):
                 If set, will return tensors of a particular framework. Acceptable values are:
-
                 - `'tf'`: Return TensorFlow `tf.constant` objects.
                 - `'pt'`: Return PyTorch `torch.Tensor` objects.
                 - `'np'`: Return NumPy `np.ndarray` objects.
                 - `'jax'`: Return JAX `jnp.ndarray` objects.
-
+            kwargs (`dict`, *optional*):
+                Keyword arguments to be passed to RobertaTokenizerFast and ClapFeatureExtractor.
```
Comment (sanchit-gandhi) [inline src/transformers/models/clap/processing_clap.py:73]: Nice docstring!

File: src/transformers/models/clap/processing_clap.py [hunk 3]
```diff
@@ -39,13 +39,20 @@ class ClapProcessor(ProcessorMixin):
...
         super().__init__(feature_extractor, tokenizer)
 
-    def __call__(self, text=None, audios=None, return_tensors=None, **kwargs):
+    def __call__(
+        self,
+        text=None,
+        audios=None,
+        return_tensors=None,
+        **kwargs,
+    ):
```
Comment (sanchit-gandhi) [inline src/transformers/models/clap/processing_clap.py:48]: Should be able to fit on one line as we had before
```suggestion
    def __call__(self, text=None, audios=None, return_tensors=None, **kwargs):
```

File: src/transformers/models/clap/processing_clap.py [hunk 4]
```diff
@@ -74,18 +85,30 @@ def __call__(self, text=None, audios=None, return_tensors=None, **kwargs):
               `None`).
             - **audio_features** -- Audio features to be fed to a model. Returned when `audios` is not `None`.
         """
-        sampling_rate = kwargs.pop("sampling_rate", None)
```
Comment (sanchit-gandhi) [inline src/transformers/models/clap/processing_clap.py:77]: This would be a breaking change: currently `sampling_rate` only gets forwarded onto the feature extractor, and never the tokenizer

Let's leave the `sampling_rate` as a special case where we don't have to prefix the `feature_extractor` arg name and it only goes to the feature extractor

**Review by sanchit-gandhi — COMMENTED** [review]
Looking really great @anmolsjoshi! Think the API here is very clean👌 Last request from me: could we add one test that shows we can correctly pass the kwargs to the feature extractor / tokenizer as required? Here, maybe you can try:
```python
feature_extractor = self.get_feature_extractor()
tokenizer = self.get_tokenizer()
processor = ClapProcessor(tokenizer=tokenizer, feature_extractor=feature_extractor)

input_audio = floats_list((3, 1000))
input_str = "This is a test string"

encoded_processor = processor(audio=input_audio, text=input_str, return_tensors="np") # everything in numpy, do some test to check this
encoded_processor = processor(audio=input_audio, text=input_str, feature_extractor_return_tensors="torch", tokenizer_return_tensors="torch") # FE in torch, tokenizer outputs in numpy, so some test to check this
```
This would be a test under: https://github.com/huggingface/transformers/blob/040c4613c2fac59f16e333a630d9a69b6ff9ca5d/tests/models/clap/test_processor_clap.py#L27

File: src/transformers/models/clap/processing_clap.py (hunk 3 above)
Comment (anmolojha) [reply src/transformers/models/clap/processing_clap.py]: Done. Changed that by mistake earlier.

File: src/transformers/models/clap/processing_clap.py (hunk 4 above)
Comment (anmolojha) [reply src/transformers/models/clap/processing_clap.py:77]: Fixed it. Added a comment saying that this is not desirable but is done for backward compatibility.

**Review by sanchit-gandhi — APPROVED** [review]
LGTM, thanks for iterating on this @anmolojha!

File: src/transformers/models/clap/processing_clap.py [hunk 5]
```diff
@@ -74,18 +81,34 @@ def __call__(self, text=None, audios=None, return_tensors=None, **kwargs):
...
+        # remove tokenizer kwargs and feature_extractor kwargs from kwargs
+        for key in tokenizer_kwargs.keys():
+            del kwargs["tokenizer_" + key]
+        for key in feature_extractor_kwargs.keys():
+            del kwargs["feature_extractor_" + key]
+
+        # collect all tokenizer kwargs and feature_extractor kwargs
+        tokenizer_kwargs = {**kwargs, **tokenizer_kwargs}
+        feature_extractor_kwargs = {"sampling_rate": sampling_rate, **kwargs, **feature_extractor_kwargs}
+
```
Comment (ylacombe) [inline src/transformers/models/clap/processing_clap.py:106]: I believe that you can use code without deleting keys from `kwargs`. Here is a code example inspired from [Bark](https://github.com/ylacombe/transformers/blob/7cd01d4e384f7ce9c18a81a4decb2c2531542661/src/transformers/models/bark/modeling_bark.py#L1520-L1544):
```suggestion
        tokenizer_kwargs = {}
        feature_extractor_kwargs = {}
        for key, value in kwargs.items():
            if key.startswith("tokenizer_"):
                key = key[len("tokenizer_") :]
                tokenizer_kwargs[key] = value
            elif key.startswith("feature_extractor_"):
                key = key[len("feature_extractor_") :]
                feature_extractor_kwargs[key] = value
            else:
                # If the key is already in a specific config, then it's been set with a
                # submodules specific value and we don't override
                if key not in tokenizer_kwargs:
                    tokenizer_kwargs[key] = value
                if key not in feature_extractor_kwargs:
                    feature_extractor_kwargs[key] = value
```

cc @amyeroberts who gave the idea for Bark !

**Review by ylacombe — APPROVED** [review]
LGTM ! I've added a code snippet proposition that I believe is cleaner, feel free to look at it !

File: src/transformers/models/clap/processing_clap.py (hunk 5 above)
Comment (anmolojha) [reply src/transformers/models/clap/processing_clap.py]: Hi @ylacombe. Yes, I agree that what you are proposing seems neater. I have modified the code to use your suggestion.

While thinking about it, a more consice way of collating all kwargs could have done:
```python
tokenizer_kwargs[key] = tokenizer_kwargs.get(key, value)
feature_extractor_kwargs[key] = feature_extractor_kwargs.get(key, value)
```
instead of this:
```python
if key not in tokenizer_kwargs:
        tokenizer_kwargs[key] = value
if key not in feature_extractor_kwargs:
        feature_extractor_kwargs[key] = value
```
But I have implemented what you suggested because I think it's easier to read.

**Comment by LysandreJik** [comment]
Would like the tokenizer expert's review cc @ArthurZucker

**Comment by sanchit-gandhi** [comment]
Gently pinging @ArthurZucker for a final review here

File: src/transformers/models/clap/processing_clap.py [hunk 6]
```diff
@@ -56,15 +57,21 @@ def __call__(self, text=None, audios=None, return_tensors=None, **kwargs):
...
+            kwargs (`dict`, *optional*):
+                Keyword arguments to be passed to RobertaTokenizerFast and ClapFeatureExtractor.
+                - Keyword arguments with `tokenizer_` prefix are passed only to RobertaTokenizerFast
+                - Keyword arguments with `feature_extractor_` prefix are passed only to ClapFeatureExtractor
+                - Keyword arguments without either of the prefix are passed to both RobertaTokenizerFast as well as
+                  ClapFeatureExtractor, **EXCEPT for `sampling_rate`**. `sampling_rate` is passed only to
+                  ClapFeatureExtractor. It is treated differently for backward compatibility and using
+                  `feature_extractor_sampling_rate` is preferred.
+                In case of a conflict, keyword arguments with prefix are given preference.
         Returns:
```
Comment (ArthurZucker) [inline src/transformers/models/clap/processing_clap.py:75]: I don't really think this is the way to go here, it's hard to maintain and would rather not have more kwargs but isolate the arguments that should be passed to the tokenizer instead of creating new arguments.

**Review by ArthurZucker — COMMENTED** [review]
Sorry for the late review.
I don't think this is aligned with our effort to get rid of kwargs, (moreover it's hard to maintain long term wise) and we should try to aim for a solution that uses explicits arguments. Isolating which arguments are need for the model at hand

**Comment by sanchit-gandhi** [comment]
Do you want to have a go at explicitly writing out the required args for the feature extractor / tokenizer @anmolojha? It should make the signature for the `__call__` method more precise as @ArthurZucker astutely mentioned.
